feat: unify SSO configuration under `.spec.sso` in backward-compatible way

jaideepr97 opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
> For example, `> /kind bug` would simply become: `/kind bug`

> /kind bug
> /kind chore
> /kind cleanup
> /kind failing-test
 /kind enhancement
> /kind documentation
> /kind code-refactoring


**What does this PR do / why we need it**:
This PR re-introduces the effort to unify existing SSO configuration options for dex and keycloak under `.spec.sso`. It intoduces 
`.spec.sso.dex` and `.spec.sso.keycloak`. It also maintains backward compatibility to support `.spec.dex`, `DISABLE_DEX` and existing `.spec.sso` fields for keycloak configuration until they are deprecated in v0.6.0. It emits events warning users of these soon-to-be-deprecated config options.
Since we need to support both the old and new config specs for dex and keycloak, this PR introduces a wide range of checks in order to ensure there is no mismatch or illegal combinations of the various SSO spec options available 

**IMPORTANT NOTES**
1. SInce https://github.com/argoproj-labs/argocd-operator/pull/615 was merged, dex health check will fail if dex is running but is not configured. Therefore this PR introduces a new constraint where dex cannot be enabled (either through `.spec.ssp.provider=dex` or `DISABLE_DEX=false`) without providing some kind of configuration in `.spec.dex` or `.spec.sso.dex`
2. if using the env var, the absence of `DISABLE_DEX` no longer implies that dex is enabled by default. It will be treated as saying dex is disabled instead. In order not to break existing workloads, if dex pods are found to be running but there is no `DISBALE_DEX` flag set, the dex resources not be deleted if any dex configuration is found to exist (either `openShiftOAuth=true` or `dex.config` is supplied). In all other cases, dex resources will be deleted

**Have you updated the necessary documentation?**

* [x] Documentation update is required by this PR.
* [ ] Documentation has been updated.

**Which issue(s) this PR fixes**:
https://github.com/argoproj-labs/argocd-operator/issues/653

Fixes https://issues.redhat.com/browse/GITOPS-1332

**How to test changes / Special notes to the reviewer**:
1. launch operator locally with `make run`
3. Create empty argo-cd instance 
```cat <<EOF | kubectl apply -f - 
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: argocd
spec: {}
EOF
```
4. Since `DISABLE_DEX` was not set, check that dex resources don't come up by default
5. Set  `.spec.sso.provider=dex` 
6. Observe dex deployment doesn't come up. Observe error in logs indicating dex must also be configured by supplying a `.spec.sso.dex` 
7. Set `.spec.sso.dex.Config="test"`
8. Observe dex deployment comes up successfully 
9. Set `.spec.sso.image=test-image`
10. Observe error in logs indicating `.spec.sso` fields (for keycloak) cannot be specified when `.spec.sso.provider=dex`
11. Remove `.spec.sso.image=test-image`. Set `.spec.sso.keycloak.image=test-image`
12. Observe error in logs indicating `.spec.sso.keycloak` cannot be specified when `.spec.sso.provider=dex`
13. Remove .spec.sso.keycloak`. Set `.spec.dex.config=test-2`
14. Observe error in logs indicating `.spec.dex` fields cannot be specified when `.spec.sso.provider=dex`
15. Remove `.spec.dex`. Remove `.spec.sso.dex`. Set `.spec.sso.provider=keycloak`
16. Observe that dex resources are deleted and keycloak resources are created successfully
17. Set `.spec.sso.image=<some-keycloak-image>` and observe it gets updated
18. Set `.spec.sso.keycloak.image=xyz`. 
19. Observe error in logs indicating that conflicting information cannot be provided in `.spec.sso` fields and equivalent `.spec.sso.keycloak` fields 
20. Remove `.spec.sso` keycloak fields (`.spec.sso.image, .spec.sso.version, .spec.sso.resources, .spec.sso.verifyTLS`) and `.spec.sso.keycloak` 
21. Set `.spec.sso.dex.config=test`
22. Observe error in logs that `.spec.sso.dex` cannot be specified when `.spec.sso.provider=keycloak`
23. Remove `.spec.sso.dex`. Set `.spec.dex.openShiftOAuth=true`
24. Observe error in logs indicating that multiple SSO providers cannot be configured simultaneously 
25. Remove `.spec.sso` entirely and observe that keycloak resources are deleted 
=====================================================================================
1. launch operator locally with `make run DISABLE_DEX=false`
2. Create empty argo-cd instance 
```cat <<EOF | kubectl apply -f - 
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: argocd
spec: {}
EOF
```
27. Notice dex pod doesn't come up. Observe error in logs indicating that dex must be configured via `.spec.dex` if dex is enabled
28. Set `.spec.dex.openShiftOAuth=true`
29. Observe that dex pod comes up successfully
30. Set `.spec.dex.openShiftOAuth=false` and `.spec.sso.provider=keycloak` 
31. Observe keycloak pod should come up in order to preserve existing behavior and not introduce a breaking change
32. Remove `.spec.sso.provider=keycloak`, Set `.spec.sso.keycloak.image=test`
33. Observe error in logs indicating that SSO provider spec cannot be supplied when `.spec.sso.provider="`
34. Remove `.spec.sso.provider` and `.spec.sso.keycloak`
35. Set `.spec.sso.image=test`
36. Observe no errors in logs
========================================================================================
37. Execute `kubectl get events`
38. Observe that deprecation events were emitted warning users that `.spec.dex`, `DISABLE_DEX` and `.spec.sso` fields for key cloak have been deprecated and support will be removed in a future release

ishitasequeira commented:
@jaideepr97 It seems the PR has a merge conflict and the e2e test needs to restart. Meanwhile, I will review the PR.

jaideepr97 commented:
@ishitasequeira updated the PR, PTAL

jopit commented:
I'm currently reviewing this, may take a while.

iam-veeramalla commented:
@jaideepr97 Can you please provide more information on how this PR is related to the point no.1 in Important Notes section.

> SInce https://github.com/argoproj-labs/argocd-operator/pull/615 was merged, dex health check will fail if dex is running but is not configured. Therefore this PR introduces a new constraint where dex cannot be enabled (either through .spec.ssp.provider=dex or DISABLE_DEX=false) without providing some kind of configuration in .spec.dex or .spec.sso.dex

jopit commented:
I've found some things that need to be fixed, but my review is ongoing.  Please don't anybody merge this before I've completed my review.

jaideepr97 commented:
@iam-veeramalla as mentioned in the description, since the health check was merged I believe dex needs to be configured if the pods are running, otherwise the health check fails 
so along with the other checks and constraints that this PR introduces, it also contains one to throw an error if someone tries to set `DISABLE_DEX=false` but doesn't configure dex through `.spec.dex.openShiftOAuth=true` or `.spec.dex.config`

This is technically a breaking change for users that currently maybe using keycloak/RHSSO but have the default `.spec.dex` and also have dex pods running in the background, but we already have broken that experience when we merged #615

jopit commented on kuttl-test.json:
```diff
@@ -0,0 +1,24 @@
+{
```
This file shouldn't be committed.  In fact, it should probably be added to .gitignore

jopit reviewed (changes requested):
Most existing Dex users will have `DISABLE_DEX` unset and will have set `.spec.Dex.Config`.  For this case the code should continue to work as it did before (i.e. it should reconcile changes), but should emit an alert telling the user they need to migrate to using `.spec.SSO`.  Instead it is explicitly leaving the existing workloads, roles etc. in place but is not reconciling.  Existing Dex users will perceive this as a major bug.  I have made more detailed review comments in the code related to this.

I also think we need two new kuttl tests, one to make sure the Dex workload, rolebinding, sa etc. are created when Dex is enabled using the new method, and the second test would be to  make sure the Dex workload, rolebinding, sa etc. are created when Dex is enabled using the old method.

jaideepr97 commented on controllers/argocd/configmap.go:
```diff
@@ -355,18 +357,26 @@ func (r *ReconcileArgoCD) reconcileArgoConfigMap(cr *argoprojv1a1.ArgoCD) error
+	// create dex config if dex is enabled either through DISABLE_DEX or through `.spec.sso`
+
+	// make sure old workloads not using DISABLE_DEX don't slip through here because their .spec.sso is nil
+	if !isDexDisabled() && isDisableDexSet && (cr.Spec.SSO == nil || cr.Spec.SSO.Provider != v1alpha1.SSOProviderTypeDex) ||
```
> If DISABLE_DEX=false and cr.Spec.SSO.Provider="dex" then this logic will skip configuring dex.
Is this a problem? I think we should be encouraging users to use one or the other

The `useDex` function sounds like a great idea, working on it now

jopit replied:
Not sure I follow.  For new users, with `DISABLE_DEX` not set, and no SSO configured (`.spec.sso = nil`), and `.spec.dex` not being set either, `useDex()` should return `false`

jopit replied:
> > If DISABLE_DEX=false and cr.Spec.SSO.Provider="dex" then this logic will skip configuring dex.
> 
> Is this a problem? I think we should be encouraging users to use one or the other

It may be a corner case, but I think users would find it very surprising that dex didn't get configured in this case.

jaideepr97 commented on controllers/argocd/dex.go:
```diff
@@ -0,0 +1,423 @@
+// reconcileDexServiceAccount will ensure that the Dex ServiceAccount is configured properly for OpenShift OAuth.
+func (r *ReconcileArgoCD) reconcileDexServiceAccount(cr *argoprojv1a1.ArgoCD) error {
+	if reflect.DeepEqual(cr.Spec.Dex, &v1alpha1.ArgoCDDexSpec{}) || !cr.Spec.Dex.OpenShiftOAuth ||
+		(cr.Spec.SSO == nil || cr.Spec.SSO.Dex == nil || !cr.Spec.SSO.Dex.OpenShiftOAuth) {
```
Ah yeah, good catch! Might have been one of the places I forgot to update the logic for that

jaideepr97 commented:
@jopit 
Thanks for the detailed review, working on addressing your concerns now

However, one aspect I'm kind of struggling with is the meaning of the `DISABLE_DEX` flag being unset 
Earlier, it meant implicit consent to have dex running, with the new way of doing things it no longer means that 

How do we support both of these interpretations simultaneously in the operator given that they are basically contradictory?
One of the original goals of this effort was to ensure that dex pods are only created when `.spec.sso.provider=dex` and that this act be de-coupled from `DISABLE_DEX`

jaideepr97 commented:
/retest

jopit reviewed (approved):
LGTM
